README
PDL::Planet

This is a package for manipulation of images, especially
pictures of Earth and satellites.

Image code taken from Xplanet by Hari Nair:

http://xplanet.sourceforge.net/

PREREQUISITES:

perl 5.6
PDL 2.4.1cvs or more recent

INSTALLATION:
 
Installation should be the normal:
 
perl Makefile.PL
make
make test
 
(as root)
make install
 
Best of luck!
 
Doug Hunt
dhunt@ucar.edu
Software Engineer III
UCAR - GPS/MET






Changes
Revision history for Perl extension PDL::Planet
0.01  01/21/2003
	- original version

0.04  03/03/2005
        - Updated build procedure to work on AMD-64.
        Must export CPPFLAGS='-fPIC' before building.
0.05  10/18/2005
        - Fixed small bug in AMD 64 detection for multi-threaded perl.

0.06  4/06/2006
        - Added new 1000x2000 earth image derived from a GMT altimetry plot
        - Added a maskColor option to the 'paste' subroutine so that
          a glyph with a black backround can be blended.
	- Added -O3 optimization to util.c.  This helps a little
          with aform_orthographic times.

0.07  5/16/2006
        - Added support for drawing the Earth's terminator:
          - the SUNPOS option to 'transform' takes the sun's
            lat/lon and shows the terminator
          - the SUNTIME option to 'transform' takes the julian time
            and uses the 'spa' package from NREL (included) to compute the sun's lat/lon
            then plotting the terminator
0.08  6/22/2006
         - Added a routine to display a day/night terminator on a linear plot.
0.09  11/16/2006
         - Added monthly earth images
0.10  3/21/2007
         - Added 3 more earth images - Maggie
0.11  1/14/2009
         - Removed -lg2c from LIBS in Makefile.PL.  This allows package to work for CentOS 5.2.
           With the change it still seems to work with CentOS 4, both 32 and 64 bit platforms.
0.12  9/9/2009
         - Added write_png function which creates an in-memory PNG in a perl scalar.  Handy
           for web applications.  Created two low-level functions write_png_mem and write_png_mem1
           which demonstrate two different ways of doing this.

test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl test.pl'

######################### We start with some black magic to print on failure.

# Change 1..1 below to 1..last_test_to_print .
# (It may become useful if the test is moved to ./t subdirectory.)

use PDL;
use PDL::Orbit;
use Test;

eval "use PDL::Planet;";
unless ($@){
  plan tests => 13;
  print "ok 1\n";
} else {
  print "not ok 1\n";
}

sub testok ($$) {
  my $bool = shift;
  my $n    = shift;
  if ($bool) {
    print "ok $n\n";
  } else {    print "not ok $n\n";
  }
}


######################### End of black magic.

BEGIN { unlink glob ("testimg*.*") };

my $pl = PDL::Planet->new
                    ->read("./images/earth.jpg")
                    ->transform(TYPE => Orthographic, SIZE => [500,500], CENTER => [40,-105])
                    ->write("testimg2.jpg");
testok -s "testimg2.jpg" > 0, 2;

my $starttime = 728784000;
my $steptimes = (sequence(50)*180) + $starttime;
my ($lat, $lon, $height) = PDL::Orbit::sattrack ($steptimes, './champ_tle.txt');

my ($x, $y)   = $pl->llh2xy($lat, $lon, $height);

my $xok = pdl(qw[463 443 416 380 339 294 246 198 151 108 70 39 16 3 0 7 24 51 86 
                 -1 -1 -1 -1 -1 -1 -1 -1 -1 503 511 509 495 471 438 397 350 299 
                 246 193 142 95 55 22 0 -13 -14 -5 13 -1 -1]);
my $yok = pdl(qw[400 430 452 467 473 471 461 442 417 385 348 308 266 223 181 143 
                 108 79 56 -1 -1 -1 -1 -1 -1 -1 -1 -1 222 264 304 343 378 407 431 
                 449 458 460 455 441 421 394 362 326 286 245 205 166 -1 -1]);

testok (sum($x - $xok) == 0, 3);
testok (sum($y - $yok) == 0, 4);

my $glyph     = PDL::Planet->new->read("./images/satellite.gif")->write("testimg5.png");
testok -s "testimg5.png" > 0, 5;

$pl->paste($glyph, $x, $y)->rgb;

my $n = 6;
foreach my $type (qw(gif jpg bmp png pbm pgm ppm tif)) {
  $pl->write("./testimg$n.$type");
  testok -s "testimg$n.$type", $n++;
}













util.c
/**----------------------------------------------------------------------
 * @name       RGB projection utilities
 *
 * @author     Doug Hunt
 * -----------------------------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "spa.h"

/**----------------------------------------------------------------------
 * @name       mag
 *
 * find the magnitude of a 3 vector
 *
 * @parameter
 * @ input:    v1 -- double array
 * @ output:   
 * @           mag -- double array
 *-----------------------------------------------------------------------*/
double mag (double v1[3]) {
  double mag;
  mag = sqrt(v1[0]*v1[0] + v1[1]*v1[1] + v1[2]*v1[2]);
  return mag;
}

/* Earth radius -- ave of max and min Earth radii, km */
#define RE 6367.44395

struct mapdefaults {
  double minlat;
  double maxlat;
  double minlon;
  double maxlon;
  double centerlon;  /* center of projection  */
  double centerlat;  /* center of projection  */
  int    minx;       /* corresponds to minlon */
  int    maxx;
  int    miny;       /* corresponds to maxlat */
  int    maxy;
  double shrink;     /* factor to shrink the Earth to show more space around , 0-1) */
  double sunlat;     /* Location of sun for computing terminator (deg) */
  double sunlon;     /* Location of sun for computing terminator (deg) */
  int    night;      /* Show night side if set to 1 */
};

static struct mapdefaults mapset = {
 -90.0,
  90.0,
-180.0,
 180.0,

   0.0, // viewer lon (deg)
  40.0, // viewer lat (deg)

     0, 
 1999,
     0,
   999,
   1.0,
  40.0,
-105.0,
 0,
};

static double pi       = 3.1415926535897932384626;
static double ncxyz[3] = {-999.0, -999.0, -999.0};

/**----------------------------------------------------------------------
 * @name       setmap
 *
 * Set static map parameters
 *
 * @parameter
 * @ input:      
 * @ maxx
 * @ maxy
 *-----------------------------------------------------------------------*/
int setmap (int maxx, int maxy) {
  mapset.maxx   = maxx;
  mapset.maxy   = maxy;
}

/**----------------------------------------------------------------------
 * @name       setcenter
 *
 * Set static map parameters
 *
 * @parameter
 * @ input:      
 * @ lat
 * @ lon
 *-----------------------------------------------------------------------*/
int setcenter (double lat, double lon) {
  double d2r  = pi/180.0;
  double cxyz[3];

  mapset.centerlat = lat;
  mapset.centerlon = lon;

  // recompute projection center normal vector
  llh2xyz (lat * d2r, lon * d2r, 0.0,  cxyz);
  norm(cxyz, ncxyz);  

}

/**----------------------------------------------------------------------
 * @name       findsunpos
 *
 * Compute the sun's subpoint given the time.
 *
 * @parameter
 * @ input:      
 * @   tjulian -- Julian time for sun direction computation
 * @ output:
 * @   lat -- degrees North, -90  to 90
 * @   lon -- degrees East,  -180 to 180
 *-----------------------------------------------------------------------*/
int findsunpos (double tjulian, double *lat, double *lon) {

  spa_data spa;
  double latwork, lonwork;
  double r2d  = 180.0/pi;
  memset(&spa, 0, sizeof(spa));

  // printf ("tjulian = %f\n", tjulian);
  // compute the Sun's inertial lat/lon and hour angle at the input Julian time
  spa.jd = tjulian;
  spa_calculate(&spa);
  // printf ("after sunpos, lat = %f, lon = %f, GAST angle = %f\n", spa.delta, spa.alpha, spa.nu);
  *lon = spa.alpha - spa.nu;
  *lat = spa.delta;

  // adjust to -180 to 180
  if (*lon >  180) *lon -= 360;
  if (*lon < -180) *lon += 360;

}

/**----------------------------------------------------------------------
 * @name       setsunpos
 *
 * Set sun position for terminator computation (also turns drawing
 * of terminator on).
 *
 * @parameter
 * @ input:      
 * @ lat -- degrees North, -90  to 90
 * @ lon -- degrees East,  -180 to 180
 *-----------------------------------------------------------------------*/
int setsunpos (double lat, double lon) {

  mapset.sunlat = lat;
  mapset.sunlon = lon;
  mapset.night  = 1;  // show night side 

}


/**----------------------------------------------------------------------
 * @name       setshrink
 *
 * Set scale factor for Earth plotting (0 to 1, 1 means Earth fills whole
 * window.
 *
 * @parameter
 * @ shrink
 *-----------------------------------------------------------------------*/
int setshrink (double shrink) {
  mapset.shrink = shrink;
}


/**----------------------------------------------------------------------
 * @name       orthographic_llh2xy
 *
 * For an orthographic projection:  convert lat/lon/height to XY plot coords
 *
 * @parameter
 * @ input:    lat -- latitude (-90 to 90)
 * @           lon -- longitude (-180(E) to 180(W))
 * @           height -- km above the Earth's surface (spherical Earth)
 * @           
 * @ output:   x -- X value of this pixel
 * @           y -- Y value of this pixel
 *-----------------------------------------------------------------------*/
int orthographic_llh2xy (double lat, double lon, double height, int *xp, int *yp) {

  double x, y;

  double d2r  = pi/180.0;
  int   ysize = mapset.maxy - mapset.miny;
  int   xsize = mapset.maxx - mapset.minx;
  double asp  = (double)ysize/(double)xsize;
  double h    = (height + RE)/RE;  // height in Earth Radii
  double midx, midy;
  double cxyz[3], xyz[3], nxyz[3];
  double center_angle, cos_theta;
  double lon0 = mapset.centerlon * d2r;
  double lat0 = mapset.centerlat * d2r; 
  double viewer_radius;  // the apparent distance from the center of the earth of the
                         // current point, as seen from infinity

  *xp = -1;  /* error value */
  *yp = -1;  /* error value */

  lat *= d2r; // convert to radians
  lon *= d2r;

  // get rid of back side points

  // compute normalized vector to input lat/lon/height
  llh2xyz (lat, lon, height, xyz);
  norm(xyz, nxyz);  

  // compute angle between viewer location and input lat/lon/height
  // note:  ncxyz is a static variable computed when 'setcenter' is called
  dot(nxyz, ncxyz, &cos_theta);
  center_angle = acos(cos_theta);

  // check for satellites that are past the horizon but high
  // enough to be visible
  viewer_radius = cos (center_angle - pi/2.0) * mag(xyz);
  
  if ((center_angle > (pi/2.0)) && (viewer_radius < (RE))) return;

  // the projection itself -- From http://mathworld.wolfram.com/OrthographicProjection.html
  // X and Y range from -1 to 1, times h
  x  = h * cos(lat) * sin(lon - lon0);
  y  = h * (cos(lat0)* sin(lat) - sin(lat0)*cos(lat)*cos(lon - lon0));

  midx = ((double)xsize-1.0)/2.0;
  midy = ((double)ysize-1.0)/2.0;

  *xp = (int)(( x * midx * asp * mapset.shrink) + midx);  
  *yp = (int)(( y * midy * mapset.shrink)       + midy); 

  return;

}

/// This routine is no longer used but is kept for reference.  See xform_orthographic1
/// below.  D. Hunt 4/19/2006
/// /**----------------------------------------------------------------------
///  * @name       xform_orthographic
///  *
///  * Convert a linear projection map into an orthographic projection
///  * map of the same size.
///  *
///  * @parameter
///  * @ input:    image -- pointer to input image
///  * @           xsize -- x dimension of image
///  * @           ysize -- y dimension of image
///  * @ output:   
///  * @           imageout -- pointer to output image (allocated by caller)
///  *-----------------------------------------------------------------------*/
/// int xform_orthographic (unsigned char *image, int xsize, int ysize,
/// 			unsigned char *imageout) {
///   int i, j;
///   int xi, yi, idx;
///   double lat, lon;
///   unsigned char this[3];
/// 
///   setmap (xsize-1, ysize-1);
///   for (j=0;j<ysize;j++) {
///     for (i=0;i<xsize;i++) {	
///       
///       lat =  ((((double)j/(double)ysize) *-180.0) + 90);
///       lon =  ((((double)i/(double)xsize) * 360.0) - 180);
///       
///       orthographic_llh2xy (lat, lon, 0.0, &xi, &yi);
///       
///       // bounds checks
///       if (xi < 0) continue;
///       if (yi < 0) continue;
///       if (xi > xsize-1) continue;
///       if (yi > ysize-1) continue;
/// 
///       this[0] = image[3*xsize*j + 3*i    ];
///       this[1] = image[3*xsize*j + 3*i + 1];
///       this[2] = image[3*xsize*j + 3*i + 2];
///       
///       idx = 3*xsize*yi + 3*xi;
///       imageout[idx  ] = this[0];
///       imageout[idx+1] = this[1];
///       imageout[idx+2] = this[2];
/// 
///       idx = 3*xsize*yi + 3*(xi+1);
///       imageout[idx  ] = this[0];
///       imageout[idx+1] = this[1];
///       imageout[idx+2] = this[2];
/// 
///       idx = 3*xsize*(yi+1) + 3*xi;
///       imageout[idx  ] = this[0];
///       imageout[idx+1] = this[1];
///       imageout[idx+2] = this[2];
/// 
///       idx = 3*xsize*(yi+1) + 3*(xi+1);
///       imageout[idx  ] = this[0];
///       imageout[idx+1] = this[1];
///       imageout[idx+2] = this[2];
/// 
///     }
///   }
/// }


/**----------------------------------------------------------------------
 * @name       xform_orthographic1
 *
 * Convert a linear projection map into an orthographic projection
 * map of the same size.  Uses a preferred reverse mapping method.
 *
 * @parameter
 * @ input:    image -- pointer to input image
 * @           xsize -- x dimension of image
 * @           ysize -- y dimension of image
 * @ output:   
 * @           imageout -- pointer to output image (allocated by caller)
 *-----------------------------------------------------------------------*/
int xform_orthographic1 (unsigned char *image, int xsize, int ysize, unsigned char *imageout) {
  int i, j;
  int xi, yi, idxout, idxin, nightTime;
  double lat, lon, x, y, c, rho, phi, lam;
  double asp  = (double)ysize/(double)xsize;
  double d2r  = pi/180.0;
  double r2d  = 180.0/pi;
  double midx = ((double)xsize-1.0)/2.0;
  double midy = ((double)ysize-1.0)/2.0;
  double lon0 = mapset.centerlon * d2r;
  double lat0 = mapset.centerlat * d2r;
  double sunECF[3];   // ECF direction of sun
  double earthECF[3]; // ECF location of current Earth point
  double costheta, theta;       // angle between these

  // Compute the XYZ ECF direction of the Sun.
  if (mapset.night) {
    llh2xyz (d2r*mapset.sunlat, d2r*mapset.sunlon, 0.0, sunECF);
    norm (sunECF, sunECF); // normalize
    // printf ("sunECF = %f, %f, %f\n", sunECF[0], sunECF[1], sunECF[2]);
  }

  for (j=0;j<ysize;j++) {
    for (i=0;i<xsize;i++) {	
      
      // scale i and j to range -1 to 1, adding in the shrink factor.
      x =      ((double)i - midx)/(midx * asp * mapset.shrink);
      y = -1.0*((double)j - midy)/(midy *       mapset.shrink);

      // printf ("y,x = %f, %f\n", y, x);

      // now do the inverse orthographic projection (from mathworld)
      rho = sqrt(x*x + y*y);
      if (rho > 1) continue; // check for other side of Earth

      c   = asin (rho);
      phi =        asin  (cos(c)*sin(lat0) + (y*sin(c)*cos(lat0))/rho);
      lam = lon0 + atan2 (x*sin(c), rho*cos(lat0)*cos(c) - y*sin(lat0)*sin(c));
      if (lam >  pi) lam -= 2.0*pi;
      if (lam < -pi) lam += 2.0*pi;

      nightTime  = 0; // Flag for night side of the Earth
      if (mapset.night) {
	//printf ("phi = %f, lam = %f\n", phi, lam);
	llh2xyz (phi, lam, 0.0, earthECF);
	norm (earthECF, earthECF); // normalize
	//printf ("earthECF = %f, %f, %f\n", earthECF[0], earthECF[1], earthECF[2]);

	// Find the angle between these vectors
	costheta = sunECF[0] * earthECF[0] +
	           sunECF[1] * earthECF[1] +
	           sunECF[2] * earthECF[2];
	//printf ("costheta = %f\n", costheta);
	// handle out of range values. 
	if (costheta >  1.0) costheta =  1.0;
	if (costheta < -1.0) costheta = -1.0;

	// angle between current point on Earth and the sun subpoint.
	// 0 at noon, pi/2 at terminator, pi at midnight
	theta = acos(costheta);
	//printf ("theta = %f\n", theta);
	if (theta > pi/2.0) nightTime  = 1;
      }

      // printf ("lat, lon = %f, %f\n", (r2d*phi), (r2d*lam));

      // convert from lat/lon radians to x/y coords 
      yi = (int)( (double)mapset.maxy * (((-r2d*phi) +  90.0)/180.0) );
      xi = (int)( (double)mapset.maxx * ((( r2d*lam) + 180.0)/360.0) );

      // printf ("yi, xi = %d, %d\n", yi, xi);

      // bounds checks
      if (xi < 0) continue;
      if (yi < 0) continue;
      if (xi > mapset.maxx) continue;
      if (yi > mapset.maxy) continue;

      idxin  = 3*mapset.maxx*yi + 3*xi;
      idxout = 3*xsize*j  + 3*i;
      if (nightTime) {
	// make night pixels dimmer
	imageout[idxout    ] = image[idxin    ]/2;
	imageout[idxout + 1] = image[idxin + 1]/2;
	imageout[idxout + 2] = image[idxin + 2]/2;
      } else {
	imageout[idxout    ] = image[idxin    ];
	imageout[idxout + 1] = image[idxin + 1];
	imageout[idxout + 2] = image[idxin + 2];
      }
    }
  }
}


/**----------------------------------------------------------------------
 * @name       apply_terminator
 *
 * Add a day/night terminator for a given time
 *
 * @parameter
 * @ input:    image -- pointer to input image
 * @ output:   
 * @           imageout -- pointer to output image (allocated by caller)
 *-----------------------------------------------------------------------*/
int apply_terminator (unsigned char *image, unsigned char *imageout) {
  int i, j, idx;
  double lat, lon;
  double d2r  = pi/180.0;
  double sunECF[3];   // ECF direction of sun
  double earthECF[3]; // ECF location of current Earth point
  double costheta, theta;       // angle between these

  // Compute the XYZ ECF direction of the Sun.
  llh2xyz (d2r*mapset.sunlat, d2r*mapset.sunlon, 0.0, sunECF);
  norm (sunECF, sunECF); // normalize
  // printf ("sunECF = %f, %f, %f\n", sunECF[0], sunECF[1], sunECF[2]);

  for (j=0;j<mapset.maxy;j++) {
    for (i=0;i<mapset.maxx;i++) {	

      // assume image runs from -180 to 180 lon (X), 90 to -90 lat (Y)
      lon = (double)i/(double)(mapset.maxx - 1);
      lon = d2r * ((lon * 360) - 180);
      lat = (double)j/(double)(mapset.maxy - 1);
      lat = -1 * d2r * ((lat * 180) - 90);

      // printf ("earthlatlon = %f, %f\n", lat, lon);
      
      llh2xyz (lat, lon, 0.0, earthECF);
      norm (earthECF, earthECF); // normalize
      //printf ("earthECF = %f, %f, %f\n", earthECF[0], earthECF[1], earthECF[2]);

      // Find the angle between these vectors
      costheta = sunECF[0] * earthECF[0] +
                 sunECF[1] * earthECF[1] +
                 sunECF[2] * earthECF[2];
      
      //printf ("costheta = %f\n", costheta);
      // handle out of range values. 
      if (costheta >  1.0) costheta =  1.0;
      if (costheta < -1.0) costheta = -1.0;

      // angle between current point on Earth and the sun subpoint.
      // 0 at noon, pi/2 at terminator, pi at midnight
      theta = acos(costheta);
      //printf ("theta = %f\n", theta);

      idx = 3*mapset.maxx*j + 3*i;

      if (theta > pi/2.0) {  // nightTime  = 1;
        // make night pixels dimmer
        imageout[idx    ] = image[idx    ]/2;
        imageout[idx + 1] = image[idx + 1]/2;
        imageout[idx + 2] = image[idx + 2]/2;
      } else {
        imageout[idx    ] = image[idx    ];
        imageout[idx + 1] = image[idx + 1];
        imageout[idx + 2] = image[idx + 2];
      }
    }
  }
}



/**----------------------------------------------------------------------
 * @name       llh2xyz
 *
 * Simple spherical to xyz coordinate conversion
 *
 * @parameter
 * @ input:    lat -- radians -pi/2 to pi/2
 * @           lon -- radians -pi to pi
 * @           height -- km
 * @ output:   
 * @           xyz -- double array, km
 *-----------------------------------------------------------------------*/
int llh2xyz (double lat, double lon, double height, double xyz[3]) {
  double colat = (pi/2.0 - lat);
  double r     = height + (RE);
  
  xyz[0] = r * cos(lon) * sin(colat);
  xyz[1] = r * sin(lon) * sin(colat);
  xyz[2] = r * cos(colat);
}

/**----------------------------------------------------------------------
 * @name       xyz2llh
 *
 * Simple xyz to spherical coordinate conversion
 *
 * @parameter
 * @ input:    xyz -- double array, km
 * @ output:   lat -- radians -pi/2 to pi/2
 * @           lon -- radians -pi to pi
 * @           height -- km
 *-----------------------------------------------------------------------*/
int xyz2llh (double xyz[3], double *lat, double *lon, double *height) {

  double phi, x, y, z, r;
  x = xyz[0];
  y = xyz[1];
  z = xyz[2];

  r     = sqrt(x*x + y*y + z*z);
  *lon   = atan2(y, x); // -pi to pi
  phi   = acos(z/r);   //   0 to pi (colatitude)

  *lat   = -(phi - pi/2);

  *height = r - (RE);

}


/**----------------------------------------------------------------------
 * @name       dot
 *
 * dot product of two 3 vectors
 *
 * @parameter
 * @ input:    v1
 * @           v2
 * @ output:   
 * @           dot -- double array
 *-----------------------------------------------------------------------*/
int dot (double v1[3], double v2[3], double *dot) {
  *dot = v1[0]*v2[0] + v1[1]*v2[1] + v1[2]*v2[2];
}

/**----------------------------------------------------------------------
 * @name       norm
 *
 * normalize a 3 vector
 *
 * @parameter
 * @ input:    v1 -- double array
 * @ output:   
 * @           norm -- double array
 *-----------------------------------------------------------------------*/
int norm (double v1[3], double norm[3]) {
  double mag;
  mag = sqrt(v1[0]*v1[0] + v1[1]*v1[1] + v1[2]*v1[2]);
  norm[0] = v1[0]/mag;
  norm[1] = v1[1]/mag;
  norm[2] = v1[2]/mag;
}


    





Makefile.PL
# Makefile.PL for a package defined by PP code.

use PDL::Core::Dev;            # Pick up development utilities
use ExtUtils::MakeMaker;
use Config;
use vars qw/$package/;

# -- Add new subroutines here! --

my @src = qw(util.c spa.c);
my @obj = @src;
map {s/\.[fc]/\.o/;} @obj; # swap .f, .c for .o

#
## Search for plplot library and include file
#

my $libdir = "./libimage";
my $incdir = "./libimage";

my %pmfiles = map { $_ => "\$(INST_LIBDIR)/Planet/$_" } glob ("images/*");
$pmfiles{'Planet.pm'} = '$(INST_LIBDIR)/Planet.pm';
#print "pmfiles = \n";
#for (keys %pmfiles) { print "$_ => $pmfiles{$_}\n"; }

$package = ["planet.pd",Planet,PDL::Planet];
WriteMakefile(
	      'NAME'  	     => 'PDL::Planet',
	      'AUTHOR'       => 'Douglas Hunt (dhunt@ucar.edu)',
	      'PREREQ_PM'    => {PDL         => 2.4.1, 
				 Image::Size => 0, 
				 PDL::Orbit  => 0},
	      'VERSION_FROM' => 'planet.pd',
	      'TYPEMAPS'     => [&PDL_TYPEMAP()],
	      'OBJECT'       => 'Planet.o ' . join (" ", @obj) . ' libimage/libimage.a',
	      'PM'           => \%pmfiles,
              'INC'          => &PDL_INCLUDE()." $plplot_include_path -I./libimage",
              'LIBS'         => ['-lgif -ljpeg -lpng -lm -lz -lpnm -ltiff -lm -lnetpbm'],
	      'clean'        => {'FILES'  => 
				   'Planet.pm Planet.xs Planet.o Planet.c Planet.bs libimage/*.o libimage/libimage.a libimage/config.* libimage/ephem libimage/Makefile'},
	      'dist'         => { COMPRESS => 'gzip', SUFFIX => 'gz' },	
	      );

sub MY::postamble { pdlpp_postamble($package); }	

# Add genpp rule; this will invoke PDL::PP on our PP file
# the argument is an array reference where the array has three string elements:
#   arg1: name of the source file that contains the PP code
#   arg2: basename of the xs and pm files to be generated  
#   arg3: name of the package that is to be generated
sub MY::postamble { 
  my $orig = pdlpp_postamble($package);

  my $cflags = '';
  # for AMD 64, use -fPIC
  if ($Config{'archname'} =~ /x86_64-linux/) {
    $cflags = '-fPIC';
  } 

  my $added = '';
  foreach my $s (@src) {
    my $o = $s;
    $o =~ s/\.[cf]/\.o/;
    
    if ($s =~ /\.c$/) { # C code
      $added .= "$o: $s\n\t$Config{'cc'} -c $cflags -o $o -I./libimage -O3 $s\n\n";
    }
    elsif ($s =~ /\.f$/) { # FORTRAN
      $added .= "$o: $s\n\tg77 -c -O3 -o $o $s\n\n";
    }
    
  }

  # compile libimage stuff
  $added .= "libimage/libimage.a: \n\tcd libimage; ./configure; make; cd ..\n\n";

  return $orig . $added;

}  


libimage/libimage.h
/****************************************************************************
    imagelib.h - header for a simple image manipulation library.
    Distributed with Xplanet.
    Copyright (C) 2002 Hari Nair <hari@alumni.caltech.edu>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
****************************************************************************/

#ifndef IMAGELIB_H
#define IMAGELIB_H

#ifdef __cplusplus
extern "C" {
#endif
    
extern int 
crop_image(unsigned char **rgb, int in_width, int in_height, int x, int y, 
           int width, int height);

extern int
delete_image(unsigned char **rgb);

extern int
read_image(const char *filename, int *width, int *height,
	   unsigned char *rgb);

extern int
resize_image(unsigned char **rgb, int in_width, int in_height, 
             int out_width, int out_height, int bilinear);

extern void
set_alpha(unsigned char *A);

extern void
set_quality(int q);

extern int
write_image(const char *filename, int width, int height, unsigned char *rgb);

extern int
write_png_mem(char **img_ptr, size_t *img_size, int width, int height, unsigned char *rgb);

#ifdef __cplusplus
}
#endif

#endif 


libimage/checkfuncs.h
/* checkfuncs.h.  Generated by configure.  */
/****************************************************************************
    checkfuncs.h - define what libraries are available.  Distributed
    with Xplanet.
    Copyright (C) 2000 Hari Nair <hari@alumni.caltech.edu>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
****************************************************************************/

#define HAVE_LIBGIF 
#define HAVE_LIBJPEG 
#define HAVE_LIBPNG 
#define HAVE_LIBPNM 
#define HAVE_LIBTIFF 


spa.h
/////////////////////////////////////////////
//          HEADER FILE for SPA.C          //
//                                         //
//      Solar Position Algorithm (SPA)     //
//                   for                   //
//        Solar Radiation Application      //
//                                         //
//               May 12, 2003              //
//                                         //
//   Filename: SPA.H                       //
//                                         //
//   Afshin Michael Andreas                //
//   afshin_andreas@nrel.gov (303)384-6383 //
//                                         //
//   Measurement & Instrumentation Team    //
//   Solar Radiation Research Laboratory   //
//   National Renewable Energy Laboratory //
//   1617 Cole Blvd, Golden, CO 80401      //
/////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////
//                                                                    //
// Usage:                                                             //
//                                                                    //
//   1) In calling program, include this header file,                 //
//      by adding this line to the top of file:                       //
//           #include "spa.h"                                         //
//                                                                    //
//   2) In calling program, declare the SPA structure:                //
//           spa_data spa;                                            //
//                                                                    //
//   3) Enter the required input values into SPA structure            //
//      (input values listed in comments below)                       //
//                                                                    //
//   4) Call the SPA calculate function and pass the SPA structure    //
//      (prototype is declared at the end of this header file):       //
//           spa_calculate(&spa);                                     //
//                                                                    //
//   Selected output values (listed in comments below) will be        //
//   computed and returned in the passed SPA structure. Output        //
//   will based on function code selected from enumeration below.     //
//                                                                    //
//   Note: A non-zero return code from spa_calculate() indicates that //
//         one of the input values did not pass simple bounds tests. //
//         The valid input ranges and return error codes are also     //
//         listed below.                                              //
//                                                                    //
////////////////////////////////////////////////////////////////////////
#ifndef __solar_position_algorithm_header
#define __solar_position_algorithm_header
//enumeration for function codes to select desired final outputs from SPA
enum {
  SPA_ZA,           //calculate zenith and azimuth
  SPA_ZA_INC,       //calculate zenith, azimuth, and incidence
  SPA_ZA_RTS,       //calculate zenith, azimuth, and sun rise/transit/set values
  SPA_ALL,          //calculate all SPA output values
};
typedef struct
{
  //----------------------INPUT VALUES------------------------
  int year;            // 4-digit year,    valid range: -2000 to 6000, error code: 1
  int month;           // 2-digit month,         valid range: 1 to 12, error code: 2
  int day;             // 2-digit day,           valid range: 1 to 31, error code: 3
  int hour;            // Observer local hour,   valid range: 0 to 24, error code: 4
  int minute;          // Observer local minute, valid range: 0 to 59, error code: 5
  int second;          // Observer local second, valid range: 0 to 59, error code: 6
  float delta_t;       // Difference between earth rotation time and terrestrial time
                       //     (from observation)
                       // valid range: -8000 to 8000 seconds, error code: 7
  float timezone;      // Observer time zone (negative west of Greenwich)
                       // valid range: -12   to   12 hours,   error code: 8

  float longitude;     // Observer longitude (negative west of Greenwich)
  // valid range: -180 to 180 degrees, error code: 9
  float latitude;      // Observer latitude (negative south of equator)
  // valid range: -90   to   90 degrees, error code: 10
  float elevation;     // Observer elevation [meters]
  // valid range: -6500000 or higher meters,    error code: 11
  float pressure;      // Annual average local pressure [millibars]
  // valid range:    0 to 5000 millibars,       error code: 12
  float temperature;   // Annual average local temperature [degrees Celsius]
  // valid range: -273 to 6000 degrees Celsius, error code; 13
  float slope;         // Surface slope (measured from the horizontal plane)
  // valid range: -360 to 360 degrees, error code: 14
  float azm_rotation; //  Surface azimuth rotation (measured from south to projection of
  //     surface normal on horizontal plane, negative west)
  // valid range: -360 to 360 degrees, error code: 15
  float atmos_refract; // Atmospheric refraction at sunrise and sunset (0.5667 deg is typical)
  // valid range: -10 to 10 degrees, error code: 16
  int function;        // Switch to choose functions for desired output (from enumeration)
  //-----------------Intermediate OUTPUT VALUES--------------------
  double jd;          //Julian day
  double jc;          //Julian century
  double jde;         //Julian ephemeris day
  double jce;         //Julian ephemeris century
  double jme;         //Julian ephemeris millennium
  double l;           //earth heliocentric longitude [degrees]
  double b;           //earth heliocentric latitude [degrees]
  double r;           //earth radius vector [Astronomical Units, AU]
  double theta;       //geocentric longitude [degrees]
  double beta;        //geocentric latitude [degrees]
  double x0;          //mean elongation (moon-sun) [degrees]
  double x1;          //mean anomaly (sun) [degrees]
  double x2;          //mean anomaly (moon) [degrees]
  double x3;          //argument latitude (moon) [degrees]
  double x4;          //ascending longitude (moon) [degrees]
  double del_psi;     //nutation longitude [degrees]
  double del_epsilon; //nutation obliquity [degrees]
  double epsilon0;    //ecliptic mean obliquity [arc seconds]
  double epsilon;     //ecliptic true obliquity [degrees]
  double del_tau;     //aberration correction [degrees]
  double lamda;       //apparent sun longitude [degrees]
  double nu0;         //Greenwich mean sidereal time [degrees]
  double nu;          //Greenwich sidereal time [degrees]
  double alpha;       //geocentric sun right ascension [degrees]
  double delta;       //geocentric sun declination [degrees]
  double h;           //observer hour angle [degrees]
  double xi;          //sun equatorial horizontal parallax [degrees]
  double del_alpha;   //sun right ascension parallax [degrees]
  double delta_prime; //topocentric sun declination [degrees]
  double alpha_prime; //topocentric sun right ascension [degrees]
  double h_prime;     //topocentric local hour angle [degrees]
  double e0;          //topocentric elevation angle (uncorrected) [degrees]
  double del_e;       //atmospheric refraction correction [degrees]
  double e;           //topocentric elevation angle (corrected) [degrees]
  double eot;         //equation of time [minutes]
  double srha;        //sunrise hour angle [degrees]
  double ssha;        //sunset hour angle [degrees]
  double sta;         //sun transit altitude [degrees]
  //---------------------Final OUTPUT VALUES------------------------
  double zenith;      //topocentric zenith angle [degrees]
  double azimuth180;  //topocentric azimuth angle (westward from south) [-180 to 180 degrees]
  double azimuth;     //topocentric azimuth angle (eastward from north) [   0 to 360 degrees]
  double incidence;   //surface incidence angle [degrees]
  double suntransit;  //local sun transit time (or solar noon) [fractional hour]
  double sunrise;     //local sunrise time (+/- 30 seconds) [fractional hour]
  double sunset;      //local sunset time (+/- 30 seconds) [fractional hour]
} spa_data;

//Calculate SPA output values (in structure) based on input values passed in structure
int spa_calculate(spa_data *spa);

//Addition D.Hunt 5/15/2006
int sunpos(spa_data *spa);
#endif

